pkg/kbagent/util/event.go

```diff
@@ -21,95 +21,142 @@ package util
 
 import (
 	"context"
+	"crypto/sha256"
+	"encoding/hex"
 	"fmt"
 	"os"
+	"sync"
 	"time"
 
+	"github.com/apecloud/kubeblocks/pkg/constant"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
 	"github.com/go-logr/logr"
-	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/rand"
 	"k8s.io/client-go/kubernetes"
 	ctlruntime "sigs.k8s.io/controller-runtime"
-
-	"github.com/apecloud/kubeblocks/pkg/constant"
 )
 
 const (
 	sendEventMaxAttempts   = 30
 	sendEventRetryInterval = 10 * time.Second
 )
 
+var (
+	baseInfo eventBaseInfo
+	once     sync.Once
+)
+
+type eventBaseInfo struct {
+	namespace string
+	podName   string
+	podUID    string
+	nodeName  string
+}
+
 func SendEventWithMessage(logger *logr.Logger, reason string, message string) {
 	go func() {
-		event := createEvent(reason, message)
-		err := sendEvent(event)
+		once.Do(func() {
+			baseInfo = eventBaseInfo{
+				namespace: os.Getenv(constant.KBEnvNamespace),
+				podName:   os.Getenv(constant.KBEnvPodName),
+				podUID:    os.Getenv(constant.KBEnvPodUID),
+				nodeName:  os.Getenv(constant.KBEnvNodeName),
+			}
+		})
+		err := sendOrUpdateEvent(reason, message)
 		if logger != nil && err != nil {
-			logger.Error(err, "send event failed")
+			logger.Error(err, "send or update event failed")
 		}
 	}()
 }
 
-func createEvent(reason string, message string) *corev1.Event {
-	// TODO(v1.0): pod variables
-	podName := os.Getenv(constant.KBEnvPodName)
-	podUID := os.Getenv(constant.KBEnvPodUID)
-	nodeName := os.Getenv(constant.KBEnvNodeName)
-	namespace := os.Getenv(constant.KBEnvNamespace)
+func sendOrUpdateEvent(reason string, message string) error {
+	suffix := hashReasonNMessage(reason, message)
+	eventName := fmt.Sprintf("%s.%s.%s", baseInfo.podName, baseInfo.podUID, suffix)
+	clientSet, err := getK8sClientSet()
+	if err != nil {
+		return fmt.Errorf("error getting k8s clientset: %v", err)
+	}
+	event, err := clientSet.CoreV1().Events(baseInfo.namespace).Get(context.TODO(), eventName, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("error getting event: %v", err)
+		}
+		event = createEvent(reason, message, eventName)
+		return sendEvent(clientSet, event)
+	}
+	return updateEvent(clientSet, event)
```

Comment:
Should there be some other conditions to determine whether to update the exist Event or create a new Event? For example, the time interval between events, or consecutive events of the same type, etc., rather than just whether the event exists or not.